go/vt/vttablet/tabletmanager/vreplication/vreplicator.go

```diff
@@ -558,3 +570,298 @@ func recalculatePKColsInfoByColumnNames(uniqueKeyColumnNames []string, colInfos
+	if err != nil {
+		return err
+	}
+	if len(secondaryKeys) > 0 {
+		alterDrop := &sqlparser.AlterTable{
+			Table: sqlparser.TableName{
+				Qualifier: sqlparser.NewIdentifierCS(vr.dbClient.DBName()),
+				Name:      sqlparser.NewIdentifierCS(tableName),
+			},
+		}
+		alterReAdd := &sqlparser.AlterTable{
+			Table: sqlparser.TableName{
+				Qualifier: sqlparser.NewIdentifierCS(vr.dbClient.DBName()),
+				Name:      sqlparser.NewIdentifierCS(tableName),
+			},
+		}
+		for _, secondaryKey := range secondaryKeys {
+			// should never happen
+			if secondaryKey.Info.Primary {
+				continue
+			}
+			alterDrop.AlterOptions = append(alterDrop.AlterOptions,
+				&sqlparser.DropKey{
+					Name: secondaryKey.Info.Name,
+					Type: sqlparser.NormalKeyType,
+				},
+			)
+			alterReAdd.AlterOptions = append(alterReAdd.AlterOptions,
+				&sqlparser.AddIndexDefinition{
+					IndexDefinition: secondaryKey,
+				},
+			)
+		}
+		action, err := json.Marshal(PostCopyAction{
+			Type: PostCopyActionSQL,
+			Task: sqlparser.String(alterReAdd),
+		})
+		if err != nil {
+			return err
+		}
+		insert, err := sqlparser.ParseAndBind(sqlCreatePostCopyAction, sqltypes.Uint32BindVariable(vr.id),
+			sqltypes.StringBindVariable(tableName), sqltypes.StringBindVariable(string(action)))
+		if err != nil {
+			return err
+		}
+		// Use a new DB client to avoid interfering with open transactions
+		// in the shared client as DDL includes an implied commit.
+		// We're also NOT using a DBA connection here because we want to
+		// be sure that the commit fails if the instance is somehow in
+		// READ-ONLY mode.
+		dbClient := vr.vre.dbClientFactoryFiltered()
+		if err := dbClient.Connect(); err != nil {
+			log.Errorf("unable to connect to the database when saving secondary keys for deferred creation: %v", err)
+			return err
+		}
+		defer dbClient.Close()
+		if _, err := dbClient.ExecuteFetch(insert, 1); err != nil {
+			return err
+		}
+		if _, err := dbClient.ExecuteFetch(sqlparser.String(alterDrop), 1); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (vr *vreplicator) getTableSecondaryKeys(ctx context.Context, tableName string) ([]*sqlparser.IndexDefinition, error) {
+	req := &tabletmanagerdatapb.GetSchemaRequest{Tables: []string{tableName}}
+	schema, err := vr.mysqld.GetSchema(ctx, vr.dbClient.DBName(), req)
+	if err != nil {
+		return nil, err
+	}
+	if len(schema.TableDefinitions) != 1 {
+		return nil, fmt.Errorf("unexpected number of table definitions returned from GetSchema call for table %q: %d",
+			tableName, len(schema.TableDefinitions))
+	}
+	ddl := schema.TableDefinitions[0].Schema
+	var secondaryKeys []*sqlparser.IndexDefinition
+	parsedDDL, err := sqlparser.ParseStrictDDL(ddl)
+	var foundCreateTable bool
+	if err != nil {
+		return secondaryKeys, err
+	}
+
+	err = sqlparser.Walk(func(node sqlparser.SQLNode) (kontinue bool, err error) {
+		switch node := node.(type) {
+		case sqlparser.DDLStatement:
+			foundCreateTable = true
```

Comment:
not sure about the terminology here. We would any `DDLStatement` imply a `foundCreateTable`?

Reply:
Yeah, I also didn't like that verbiage but forgot to change it. Will do so now.